scala/src/com/google/idea/blaze/scala/run/producers/ScalaTestContextProvider.java

```diff
@@ -76,32 +92,49 @@ private static ScClass getTestClass(ConfigurationContext context) {
       // handled by a different producer
       return null;
     }
-    return getTestClass(location);
+    ScalaTestConfigurationProducer scalaTestConfigurationProducer = ScalaTestConfigurationProducer.instance();
+    return scalaTestConfigurationProducer.getTestClassWithTestName(location)
+        .map(classWithTestName -> new ClassAndTestName(classWithTestName.testClass(), classWithTestName.testName().getOrElse(() -> null)))
+        .getOrElse(() -> null);
   }
 
-  @Nullable
-  private static ScClass getTestClass(Location<?> location) {
-    PsiElement element = location.getPsiElement();
-    ScClass testClass;
-    if (element instanceof ScClass) {
-      testClass = (ScClass) element;
-    } else {
-      testClass = PsiTreeUtil.getParentOfType(element, ScClass.class);
-    }
-    if (testClass != null && isTestClass(testClass)) {
-      return testClass;
-    }
-    return null;
-  }
+  static TestContext.BlazeFlagsModification setScalatestTestSelectorArguments(ClassAndTestName classAndTestName) {
+    String testClassSelectorFlag = BlazeFlags.TEST_ARG + "-s";
+    String testClassFlag = BlazeFlags.TEST_ARG + classAndTestName.testClass.getQualifiedName();
+    String testNameSelectorFlag = BlazeFlags.TEST_ARG + "-t";
+    // Scalatest names can contain spaces, so the name needs to be quoted
+    String testNameFlag = BlazeFlags.TEST_ARG + "\"" + classAndTestName.testName + "\"";
```

Comment:
We should also escape some characters here, at least `"` and `\`

Reply:
I understand the need for `"`, but why is escaping `\` needed?

Reply:
That's tricky, but this fails when you try to run this
```
class Test2 extends AnyFunSuite {
  test("Test\\") {

  }
}
```

In order to fix it I had to manually edit configuration and set it to 
```
--test_arg=-s
--test_arg=foo.Test2
--test_arg=-t
--test_arg=Test\
```

Reply:
:+1: Tried to address this, and added a test for it.

Reply:
Having dug into this, I think the backslash issue isn't solvable with changes in just this code.

We postprocess the flags in RunConfigurationFlagsState
```
flags.stream()
    .map(s -> ParametersListUtil.parse(s, false, true).get(0))
    .collect(Collectors.toList());
```

I ran a few tests on this util, and I'm not sure these lines do what they're supposed to.

`parse("--test_arg=this is a string").get(0)` gets you `--test_arg=this`, with anything after the whitespace stripped. 
I think this is why I have to add quotes around the test name, or we lose most of the name here.

`parse("--test_arg=\"this is a string\"").get(0)` gets you `--test_arg=this is a string`. 
Notice that this string doesn't contain quotes yet. Something later in the process must handle adding quotes for flags containing whitespace.

`parse("--test_arg="this is a string\\\"")` yields `--test_arg=this is a string"`. 
`parse("--test_arg="this is a string\\\\\"")` yields `--test_arg=this is a string\"`
I don't see a way to deal with a trailing backslash when the strings pass through this method, since any backslashes here get interpreted as an escape for the following quote.

I don't really understand why the postprocessing code I posted above throws away most of the parameter list, but adding quotes on my end doesn't seem like the right solution. Maybe RunConfigurationFlagsState should keep the entire flag including whitespaces, since it seems like something later on is adding quotes if necessary?

I'd prefer not dealing with this issue here. I think the inability to deal with backslashes could be tracked as a separate issue. I don't think it's something I can fix locally in the flags modifier, and I don't think backslashes in testnames are so crucial they should block merging this.

What do you think?

Reply:
I've removed the backslash replacement and the misleading test, since it doesn't work.

Reply:
Resolving this, since it is a likely rare edge case that seems to require a decent amount of work to fix. Even if this is worth fixing, I don't think it should block the rest of this code from going in, since it is not a regression.